xtable-utilities/src/main/java/org/apache/xtable/utilities/RunCatalogSync.java

```diff
@@ -0,0 +1,381 @@
+    Configuration hadoopConf = loadHadoopConf(customConfig);
+
+    customConfig = getCustomConfigurations(cmd, CONVERTERS_CONFIG_PATH);
+    TableFormatConverters tableFormatConverters = loadTableFormatConversionConfigs(customConfig);
+
+    Map<String, ExternalCatalogConfig> catalogsById =
+        datasetConfig.getTargetCatalogs().stream()
+            .map(RunCatalogSync::populateCatalogImplementations)
+            .collect(Collectors.toMap(ExternalCatalogConfig::getCatalogId, Function.identity()));
+    Optional<CatalogConversionSource> catalogConversionSource =
+        getCatalogConversionSource(datasetConfig.getSourceCatalog(), hadoopConf);
+    ConversionController conversionController = new ConversionController(hadoopConf);
+    for (DatasetConfig.Dataset dataset : datasetConfig.getDatasets()) {
+      SourceTable sourceTable =
+          getSourceTable(dataset.getSourceCatalogTableIdentifier(), catalogConversionSource);
+      List<TargetTable> targetTables = new ArrayList<>();
+      Map<TargetTable, List<TargetCatalogConfig>> targetCatalogs = new HashMap<>();
+      for (TargetTableIdentifier targetCatalogTableIdentifier :
+          dataset.getTargetCatalogTableIdentifiers()) {
+        TargetTable targetTable =
+            TargetTable.builder()
+                .name(sourceTable.getName())
+                .basePath(sourceTable.getBasePath())
+                .namespace(sourceTable.getNamespace())
+                .formatName(targetCatalogTableIdentifier.getTableFormat())
+                .build();
+        targetTables.add(targetTable);
+        if (!targetCatalogs.containsKey(targetTable)) {
+          targetCatalogs.put(targetTable, new ArrayList<>());
+        }
+        targetCatalogs
+            .get(targetTable)
+            .add(
+                TargetCatalogConfig.builder()
+                    .catalogTableIdentifier(
+                        getCatalogTableIdentifier(
+                            targetCatalogTableIdentifier.getTableIdentifier()))
+                    .catalogConfig(catalogsById.get(targetCatalogTableIdentifier.getCatalogId()))
+                    .build());
+      }
+      ConversionConfig conversionConfig =
+          ConversionConfig.builder()
+              .sourceTable(sourceTable)
+              .targetTables(targetTables)
+              .targetCatalogs(targetCatalogs)
+              .syncMode(SyncMode.INCREMENTAL)
+              .build();
+      List<String> tableFormats =
+          new ArrayList<>(Collections.singleton(sourceTable.getFormatName()));
+      tableFormats.addAll(
+          targetTables.stream().map(TargetTable::getFormatName).collect(Collectors.toList()));
+      tableFormats = tableFormats.stream().distinct().collect(Collectors.toList());
+      try {
+        conversionController.syncTableAcrossCatalogs(
+            conversionConfig,
+            getConversionSourceProviders(tableFormats, tableFormatConverters, hadoopConf));
```

Comment:
I found this a bit confusing when reading through the code since it is using targets to create sources at first glance. Would the `SourceProvider` be more clear if it was a `TableStateProvider` or something that does not include source/target language we use elsewhere for the table format sync?

Reply:
I will think of a better name, for now keeping this refactoring in a separate PR. 
https://github.com/apache/incubator-xtable/pull/613